packages/composable-controller/CHANGELOG.md

```diff
@@ -7,6 +7,16 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+### Changed
+
+- **BREAKING:** `ComposableController` constructor option `controllers` and generic type argument `ChildControllers` are re-defined from an array of controller instances to an object that maps controller names to controller instances.
+- **BREAKING:** `ComposableController` class field objects `state` and `metadata` exclude child controllers that do not extend from `BaseController` or `BaseControllerV1`. Any non-controller entries that are passed into the constructor will be removed automatically.
+
+### Fixed
+
+- **BREAKING:** `ComposableController` class field object `metadata` now assigns the `StateMetadataProperty`-type object `{ persist: true, anonymous: true }` to each child controller name.
```

Comment:
Sorry if I missed something, but why do we want to do this now? Why are we assuming that all state for a controller is anonymous?

Reply:
This fixes a clear error. Right now we have this:

```
type ComposableControllerStateMetadata = { 
  ChildController1: { 
    childControllerProperty1: { persist: boolean; anonymous: boolean };, 
    childControllerProperty2: { persist: boolean; anonymous: boolean };
    ...
  };
  ChildController2: { ... };
  ...
}
```

When what `ComposableController` as a V2 controller requires is this:

```
type ComposableControllerStateMetadata = {
  ChildController1: { persist: boolean; anonymous: boolean };
  ChildController2: { persist: boolean; anonymous: boolean };
  ...
}
```

Setting `anonymous` as 'true' just seemed like the safer option, but I can change that if we want to make sure that all child controller state properties in `ComposableController` are not obfuscated.

Reply:
Using `anonymous: true` already ensures that they are not obfuscated by assuming that they contain no PII, wouldn't it? That's why I raised this question — it seems like may not want to do that in all cases. However 1) it's possible I am misunderstanding how ComposableController state will eventually get used and 2) I see we were already using `anonymous: true` for for BaseController v1 controllers or non-controllers. So, minimally it _seems_ that we would want to reuse the controller metadata in all cases if we can, since we've already made a decision on the controller level about whether certain properties have PII or not. But then again there may be something I am missing here...

Reply:
Even if `anonymous` were set to false, would the obfuscation be correctly applied to an entire state object? Since other controller instances in the client would have access to `ComposableController` state via `getState` action, we should probably prevent PII from being made available.

Reply:
Hmm... I thought I understood how the metadata gets used, but after looking around, I am now unsure. I know we have a `getAnonymizedState` function in `@metamask/base-controller`, but it doesn't seem to be getting used in the extension 🤔

Reply:
Based on [this file](https://github.com/MetaMask/metamask-extension/blob/589d7e4d1b45723183b825ba958d482fad53a23b/app/scripts/constants/sentry-state.ts), it seems that the extension maintains its own list of which state properties are anonymous or not instead of consulting the controller metadata (and [here](https://github.com/MetaMask/metamask-extension/blob/589d7e4d1b45723183b825ba958d482fad53a23b/shared/modules/object.utils.js#L23) is where we do the obfuscation). It looks like we are doing a similar thing for mobile [here](https://github.com/MetaMask/metamask-mobile/blob/1a0dc3b14c03c8a62b3fe3c0a9ef6d67dcb3234f/app/util/sentry/utils.js).

I suppose we can go with your changes for now, and once we start using the controller metadata in clients, we will have a more informed decision we can make on how best to set the metadata on the ComposableController level.